src/mlpack/methods/ann/visitor/output_size_visitor.hpp

```diff
@@ -0,0 +1,75 @@
+/**
+ * @file output_size_visitor.hpp
+ * @author Atharva Khandait
+ *
+ * This file provides an abstraction for the OutputSize() function for
+ * different layers and automatically directs any parameter to the right layer
+ * type.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_VISITOR_OUTPUT_SIZE_VISITOR_HPP
+#define MLPACK_METHODS_ANN_VISITOR_OUTPUT_SIZE_VISITOR_HPP
+
+#include <mlpack/methods/ann/layer/layer_traits.hpp>
+
+#include <boost/variant.hpp>
+
+namespace mlpack {
+namespace ann {
+
+/**
+ * OutputWidthVisitor exposes the OutputSize() method of the given module.
+ */
+class OutputSizeVisitor : public boost::static_visitor<size_t>
+{
+ public:
+  //! Return the output size.
+  template<typename LayerType>
+  size_t operator()(LayerType* layer) const;
+
+  size_t operator()(MoreTypes layer) const;
+
+ private:
+  //! Return 0 if the module doesn't implement the InputSize() or Model()
+  //! function.
+  template<typename T>
+  typename std::enable_if<
+      !HasInputSize<T, size_t&(T::*)()>::value &&
```

Comment:
I'm pretty sure that the builds are failing due to an incorrect function pointer here and the next four functions. The `Linear()` object has an `InputSize()` function which means that the second function should be called here. However, the first one is getting called and that's probably why the test is failing. I had written these for different function types when #1985 wasn't being worked on. Can someone help me with the correct function pointer here. I tried `size_t(T::*)()` but it's not working.

Reply:
Could this be because the method is `const`?  Maybe try this signature: `size_t(T::*)() const`... I think that one will work; not sure though.

Reply:
Hey @akhandait, have you had a chance to try that fix yet?  It would be great to get this merged, and it's so close. :)

Reply:
Hey, I'm really sorry about taking so long. I have been very busy with college work. I had tried what you suggested but it didn't quite work. I'll try it again tomorrow night and get back. This PR has been around for too long now.

Reply:
No worries!  I know how busy things get.  If you don't get to it, it's okay, someone else can handle it too; just let us know if you want a hand. :)

Reply:
Thanks for understanding! I'll give it one more try and then let you know if I need any help. :)

Reply:
It was a very simple fix, should have figured it out a month ago. However, I had to remove the const from `InputSize()` and `OutputSize()` because I am changing the values through the visitor.

Reply:
Sounds good---awesome to see it worked out. :)